src/frontends/onnx/frontend/src/editor.cpp

```diff
@@ -175,6 +175,90 @@ void modify_initializer(TensorProto& initializer,
     }
 }
+
+bool is_topologically_sorted(const GraphProto& graph) {
+    std::unordered_set<std::string> known_tensors;
+    std::transform(std::begin(graph.input()),
+                   std::end(graph.input()),
+                   std::inserter(known_tensors, std::end(known_tensors)),
+                   extract_name<ONNX_NAMESPACE::ValueInfoProto>);
+    std::transform(std::begin(graph.output()),
+                   std::end(graph.output()),
+                   std::inserter(known_tensors, std::end(known_tensors)),
+                   extract_name<ONNX_NAMESPACE::ValueInfoProto>);
+    std::transform(std::begin(graph.initializer()),
+                   std::end(graph.initializer()),
+                   std::inserter(known_tensors, std::end(known_tensors)),
+                   extract_name<ONNX_NAMESPACE::TensorProto>);
+
+    for (const auto& node : graph.node()) {
+        for (const auto& input : node.input()) {
+            if (input.empty()) {
+                continue;
+            }
+            if (known_tensors.count(input) == 0) {
+                return false;
+            }
+        }
+        for (const auto& output_name : node.output()) {
+            known_tensors.insert(output_name);
+        }
+    }
+    return true;
+}
+
+void graph_topological_sort(GraphProto* graph) {
+    if (!is_topologically_sorted(*graph)) {
+        std::stack<const NodeProto*, std::vector<const NodeProto*>> nodes_to_process;
+        std::unordered_set<const NodeProto*> processed_nodes;
+        std::multimap<std::string, const NodeProto*> output_name_to_node;
+        GraphProto result;
+
+        for (size_t i = 0; i < graph->node().size(); ++i) {
+            for (const auto& output_name : graph->node(i).output()) {
+                output_name_to_node.emplace(output_name, graph->mutable_node(static_cast<int>(i)));
+            }
+        }
+        auto get_node_by_out_name = [&output_name_to_node](const std::string& out_name) -> const NodeProto* {
+            const auto& idx_iter = output_name_to_node.find(out_name);
+            if (idx_iter != std::end(output_name_to_node)) {
+                return idx_iter->second;
+            }
+            return nullptr;
+        };
+        for (const auto& node : graph->node()) {
+            nodes_to_process.push(&node);
+        }
+
+        while (nodes_to_process.size() > 0) {
+            auto* node = nodes_to_process.top();
+            if (processed_nodes.count(node) == 0) {
+                bool can_add = true;
+                for (const auto& in_name : node->input()) {
+                    const auto* in_node = get_node_by_out_name(in_name);
+                    if (in_node == nullptr) {  // can be an initializer or an model's input
+                        continue;
+                    }
+                    if (processed_nodes.count(in_node) == 0) {
+                        can_add = false;
+                        nodes_to_process.push(in_node);
+                    }
```

Comment:
Pushing nodes here is needed (to traverse the graph up), but you are right, that checking on the beginning of the while loop is not needed. Thank you!

Reply:
> Just wondering, could it be reduced for example by using model outputs names as initial nodes to process.
Thank you for this comment. I've checked it at the begining of working on this task and measured time using models with fully reversed nodes order. The time was shorter in version where `nodes_to_process` is initialized with all nodes. But I've checked again using models where only a few nodes are unsorted and you are right, that for such cases using version where only nodes producing outputs are pushed into `nodes_to_process` is faster. Because it's look like a more real case, I've applied your suggestion ;)